### ISSUE TITLE
abstract classes with initialize get Ruby --verbose warnings

### ISSUE THREAD

**Issue by md-work** [sorbet/sorbet#2374]
Classes marked with `abstract!` may contain an `initialize`. This is because:
- subclasses should be able to call it via super
- the class itself may want to call the `initialize` of it's parent class (and the parent may not be abstract)
&nbsp;

Sorbet-Runtime defines a method `initialize` when `abstract!` is called. This first `initialize` will then be overwritten by the individual `initialize`. And such an overwriting of methods will be marked as warning when using Ruby with `$VERBOSE = true`.
&nbsp;

Example:
```ruby
$ cat exmaple.rb
require 'sorbet-runtime'

class Example
  extend T::Helpers
  abstract!

  def initialize; end
end
$ ruby --verbose exmaple.rb 
exmaple.rb:7: warning: method redefined; discarding old initialize
/usr/lib64/ruby/gems/2.5.0/gems/sorbet-runtime-0.4.5072/lib/types/private/abstract/declare.rb:37: warning: previous definition of initialize was here
```
&nbsp;

Workaround:
Create an alias for `initialize`, so Ruby --verbose won't mind.
```ruby
  alias _initialize initialize  # TODO: workaround ruby--verbose warning for sorbet-runtime 
  def initialize; end
```

**Comment by jez** [comment]
I would not recommend this workaround. Sorbet defines an `initialize` method in the abstract class to prevent against people actually instantiating that class at runtime. By using this `alias` thing, you're throwing away that sanity check.

**Comment by md-work** [comment]
> I would not recommend this workaround. Sorbet defines an `initialize` method in the abstract class to prevent against people actually instantiating that class at runtime. By using this `alias` thing, you're throwing away that sanity check.

That's why it's just a workaround.
**I'm still hoping for a better solution by Sorbet!**

&nbsp;

---

&nbsp;

Nevertheless, I had some thoughts about improving the workaround.
What if I call that aliased initialize as first instruction of my new custom initialize method?

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%24VERBOSE%20%3D%20true%0Arequire%20'sorbet-runtime'%0A%0Aclass%20Abstract%0A%20%20extend%20T%3A%3AHelpers%0A%20%20abstract!%0A%20%20%0A%20%20alias%20initialize_abstract%20initialize%20%20%23%20TODO%3A%20workaround%20ruby--verbose%20warning%20for%20sorbet-runtime%20%0A%20%20def%20initialize%0A%20%20%20%20initialize_abstract%0A%20%20%20%20puts('Abstract')%0A%20%20end%0Aend%0A%0Aclass%20Concrete%20%3C%20Abstract%0A%20%20def%20initialize%0A%20%20%20%20super%0A%20%20end%0Aend%0A%0Aputs(Concrete.new.class)%0A%23%20%3D%3E%0A%23%20%20Abstract%0A%23%20%20Concrete%0A%0Aputs(Abstract.new.class)%0A%23%20%3D%3E%0A%23%20%20Abstract%20is%20declared%20as%20abstract%3B%20it%20cannot%20be%20instantiated%20(RuntimeError))
```ruby
# typed: true
$VERBOSE = true
require 'sorbet-runtime'

class Abstract
  extend T::Helpers
  abstract!
  
  alias initialize_abstract initialize  # TODO: workaround ruby--verbose warning for sorbet-runtime 
  def initialize
    initialize_abstract
    puts('Abstract')
  end
end

class Concrete < Abstract
  def initialize
    super
  end
end

puts(Concrete.new.class)
# =>
#  Abstract
#  Concrete

puts(Abstract.new.class)
# =>
#  Abstract is declared as abstract; it cannot be instantiated (RuntimeError)
```

**Comment by will** [comment]
@md-work thanks for the workaround! It's a great and very necessary workaround, and I'm using it, and will continue to use it until there is some real solution in sorbet.

**Comment by marcandre** [comment]
I'm proposing a solution in #3388

**Comment by jez** [comment]
The question of the warning being emitted was fixed in #4695.

#3388 can remain open for the question of `new` vs `initialize`.
